File: src/hittable/sphere.rs
use crate::vector::*;

use super::HitRecord;
use super::Hittable;

pub struct Sphere {
    center: Point3,
    radius: f64,
}

impl Sphere {
    pub fn new(cen: &Point3, rad: f64) -> Self {
        Self {
            center: *cen,
            radius: rad,
        }
    }
}

impl Hittable for Sphere {
    fn hit(&self, r: &crate::ray::Ray, t_min: f64, t_max: f64, rec: &mut HitRecord) -> bool {
        let diff = r.origin() - self.center;
        let a = r.direction().length_squared();
        let half_b = dot(&r.direction(), &diff);
        let c = diff.length_squared() - self.radius * self.radius;

        let discriminant = half_b * half_b - a * c;
        if discriminant < 0.0 {
            return false;
        }

        let sqrtd = discriminant.sqrt();
        // calculate root between t_min, t_max
        let mut root = (-half_b - sqrtd) / a;
        if root < t_min || t_max < root {
            root = (-half_b + sqrtd) / a;
            if root < t_min || t_max < root {
                return false;
            }
        }

        rec.t = root;
        rec.p = r.at(rec.t);
        let outward_normal = (rec.p - self.center) / self.radius;
        rec.set_face_normal(r, &outward_normal);

        true
    }
}

File: CPU/src/geometry/sphere.cpp
#include "geometry/sphere.h"
#include "materials/material.h"
#include "math/ray.h"
#include <cmath>

Sphere::Sphere(const Point3& center, float radius, std::shared_ptr<Material> material)
    : center(center), radius(radius), material(material) {}

bool Sphere::hit(const Ray& ray, float t_min, float t_max, HitRecord& rec) const {
    Vec3 oc = ray.origin - center;
    float a = ray.direction.length_squared();
    float half_b = oc.dot(ray.direction);
    float c = oc.length_squared() - radius * radius;
    
    float discriminant = half_b * half_b - a * c;
    if (discriminant < 0) return false;
    
    float sqrtd = sqrt(discriminant);
    
    // Find the nearest root that lies in the acceptable range
    float root = (-half_b - sqrtd) / a;
    if (root < t_min || t_max < root) {
        root = (-half_b + sqrtd) / a;
        if (root < t_min || t_max < root)
            return false;
    }
    
    rec.t = root;
    rec.point = ray.at(rec.t);
    Vec3 outward_normal = (rec.point - center) / radius;
    rec.set_face_normal(ray, outward_normal);
    rec.material = material.get();
    
    return true;
}

BoundingBox Sphere::bounding_box() const {
    Vec3 radius_vec(radius, radius, radius);
    return BoundingBox(center - radius_vec, center + radius_vec);
}